src/cyclotomic.rs

```diff
@@ -97,80 +99,121 @@ impl Cyclotomic {
-        .map(|k| new_order * k / z.order)
-        .collect();
-
-    z.order = new_order
+    fn is_zero(&self) -> bool {
+        return self.order == 1
+            && self.coeffs == vec![Q::new(Z::from(0), Z::from(1))]
+            && self.exps == vec![];
+    }
 }
 
-// TODO: Use Rob's code to actually do some reductions here?
-pub fn match_orders(z1: &mut Cyclotomic, z2: &mut Cyclotomic) -> () {
-    let new_order = num::integer::lcm(z1.order, z2.order);
-    increase_order_to(z1, new_order);
-    increase_order_to(z2, new_order);
+impl One for Cyclotomic {
+    fn one() -> Self {
+        Cyclotomic::new(1, vec![Q::new(Z::from(1), Z::from(1))], vec![0])
+    }
 }
 
-// Note: checking equality mutates the encoding of $z_1$ and $z_2$, but does not
-// change the values represented. Since we mutate, we cannot use the Rust traits
-// \tt{Eq} and \tt{PartialEq}. The same is true for all arithmetic operations.
-pub fn eq(z1: &mut Cyclotomic, z2: &mut Cyclotomic) -> bool {
-    match_orders(z1, z2);
-    // Since we assume the exponent list is sorted, we can just do a
-    // comparison of the coeff and exp vectors, the number is in
-    // canonical form.
-    z1.coeffs == z2.coeffs && z1.exps == z2.exps
+impl Cyclotomic {
+    // TODO: right now we INCREASE the orders to get compatibility... we should try to
+    // DECREASE order first!
+    pub fn increase_order_to(z: &mut Cyclotomic, new_order: u64) -> () {
+        z.exps = z
+            .exps
+            .clone()
+            .into_iter()
+            .map(|k| new_order * k / z.order)
+            .collect();
+
+        z.order = new_order
+    }
+
+    // TODO: Use Rob's code to actually do some reductions here?
+    pub fn match_orders(z1: &mut Cyclotomic, z2: &mut Cyclotomic) -> () {
+        let new_order = num::integer::lcm(z1.order, z2.order);
+        Cyclotomic::increase_order_to(z1, new_order);
+        Cyclotomic::increase_order_to(z2, new_order);
+    }
 }
 
-pub fn add(z1: &mut Cyclotomic, z2: &mut Cyclotomic) -> Cyclotomic {
-    match_orders(z1, z2);
-
-    let mut z1_i = 0;
-    let mut z2_i = 0;
-    let mut result = zero();
-
-    // We are producing a result cyclotomic in canonical form assuming the
-    // inputs are in canonical form. That is, we build the result term by
-    // term, assuming that the exponent lists are sorted.
-    while z1_i < z1.exps.len() || z2_i < z2.exps.len() {
-        while z1_i < z1.exps.len() && z1.exps[z1_i] < z2.exps[z2_i] {
-            result.coeffs.push(z1.coeffs[z1_i].clone());
-            result.exps.push(z1.exps[z1_i].clone());
-            z1_i += 1;
-        }
+impl PartialEq for Cyclotomic {
+    // Note: checking equality mutates the encoding of $z_1$ and $z_2$, but does not
+    // change the values represented. Since we mutate, we cannot use the Rust traits
+    // \tt{Eq} and \tt{PartialEq}. The same is true for all arithmetic operations.
+    fn eq(&self, other: &Self) -> bool {
+        let mut z1 = self.clone();
```

Comment:
This is fine for convenience for now, but it's obvious that this copy is done for no reason. Yes, checking equality shouldn't mutate, but it also shouldn't use any extra space. And the `match_orders` mutation isn't a real mutation.